src/Storages/StorageFile.cpp

```diff
@@ -526,12 +495,182 @@ namespace
+            , archive_info(archive_info_)
+            , format(format_)
+            , format_settings(format_settings_)
+        {
+        }
+
+        std::unique_ptr<ReadBuffer> next() override
+        {
+            std::unique_ptr<ReadBuffer> read_buf;
+            struct stat file_stat;
+            while (true)
+            {
+                if (current_archive_index == archive_info.paths_to_archives.size())
+                {
+                    if (is_first)
+                        throw Exception(
+                            ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE,
+                            "Cannot extract table structure from {} format file, because all files are empty. You must specify table structure manually",
+                            format);
+
+                    return nullptr;
+                }
+
+                const auto & archive = archive_info.paths_to_archives[current_archive_index];
+                file_stat = getFileStat(archive, false, -1, "File");
+                if (file_stat.st_size == 0)
+                {
+                    if (getContext()->getSettingsRef().engine_file_skip_empty_files)
+                    {
+                        ++current_archive_index;
+                        continue;
+                    }
+
+                    throw Exception(
+                        ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE,
+                        "Cannot extract table structure from {} format file, because the archive {} is empty. "
+                        "You must specify table structure manually",
+                        format,
+                        archive);
+                }
+
+                auto archive_reader = createArchiveReader(archive);
+
+                auto try_get_columns_from_schema_cache = [&](const std::string & full_path) -> std::optional<ColumnsDescription>
+                {
+                    auto context = getContext();
+                    if (!getContext()->getSettingsRef().schema_inference_use_cache_for_file)
+                        return std::nullopt;
+
+                    auto & schema_cache = StorageFile::getSchemaCache(context);
+                    auto get_last_mod_time = [&]() -> std::optional<time_t>
+                    {
+                        if (0 != stat(archive_reader->getPath().c_str(), &file_stat))
+                            return std::nullopt;
+
+                        return file_stat.st_mtime;
+                    };
+
+                    auto cache_key = getKeyForSchemaCache(full_path, format, format_settings, context);
+                    auto columns = schema_cache.tryGetColumns(cache_key, get_last_mod_time);
+
+                    if (columns)
+                        return columns;
+
+                    return std::nullopt;
+                };
+
+                if (archive_info.isSingleFileRead())
+                {
+                    read_buf = archive_reader->readFile(archive_info.path_in_archive, false);
```

Comment:
What if you read from multiple archives, and some of those archives don't have a file you requested?
I think this behavior should be controlled with a setting but in general, I left it to just skip because it would be annoying to have the entire reading fail because of 1 archive who is missing the requested file.